LibIPC: Error when forgetting to name parameter is inscrutable

If you add an IPC method like so:

```
some_method(MyNS::MyCoolType) =|
```

Then the IPCCompiler does not recognize that this is an error. Parameters must be named.

Instead, it gives the parameter a type of `MyNS::MyCoolType)` and a name of `=|` and produces all sorts of malformed code.

e.g.

```console
/Users/andrew/Source/serenity/Build/lagom/Userland/Services/WebContent/WebContentClientEndpoint.h:128:55: error: expected expression
  128 |     NotifyProcessInformation(WebView::ProcessHandle) =|, did_start_loading(u64 page_id, URL::URL url, bool is_redirect) : m_=|(move(=|)), m_page_id(move(page_id)), m_url(move(url)), m_is_redirect(move(is_redirect)) {}
      |                                                       ^
/Users/andrew/Source/serenity/Build/lagom/Userland/Services/WebContent/WebContentClientEndpoint.h:128:56: error: expected expression
  128 |     NotifyProcessInformation(WebView::ProcessHandle) =|, did_start_loading(u64 page_id, URL::URL url, bool is_redirect) : m_=|(move(=|)), m_page_id(move(page_id)), m_url(move(url)), m_is_redirect(move(is_redirect)) {}
      |                                                        ^
/Users/andrew/Source/serenity/Build/lagom/Userland/Services/WebContent/WebContentClientEndpoint.h:128:58: error: a type specifier is required for all declarations
  128 |     NotifyProcessInformation(WebView::ProcessHandle) =|, did_start_loading(u64 page_id, URL::URL url, bool is_redirect) : m_=|(move(=|)), m_page_id(move(page_id)), m_url(move(url)), m_is_redirect(move(is_redirect)) {}
      |                                                          ^
/Users/andrew/Source/serenity/Build/lagom/Userland/Services/WebContent/WebContentClientEndpoint.h:128:121: error: expected ';' at end of declaration list
  128 |     NotifyProcessInformation(WebView::ProcessHandle) =|, did_start_loading(u64 page_id, URL::URL url, bool is_redirect) : m_=|(move(=|)), m_page_id(move(page_id)), m_url(move(url)), m_is_redirect(move(is_redirect)) {}
      |                                                                                                                         ^
      |                                                                                                                         ;
```

--- a/Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp
+++ b/Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp
@@ -140,8 +140,8 @@ Vector<Endpoint> parse(ByteBuffer const& file_contents)
         return parameter_type;
     };
 
-    auto parse_parameter = [&](Vector<Parameter>& storage) {
-        for (;;) {
+    auto parse_parameter = [&](Vector<Parameter>& storage, StringView message_name) {
+        for (auto parameter_index = 1;; ++parameter_index) {
             Parameter parameter;
             if (lexer.is_eof()) {
                 warnln("EOF when parsing parameter");
@@ -165,6 +165,10 @@ Vector<Endpoint> parse(ByteBuffer const& file_contents)
                 }
             }
             parameter.type = parse_parameter_type();
+            if (parameter.type.ends_with(',') || parameter.type.ends_with(')')) {
+                warnln("Parameter {} of method: {} must be named", parameter_index, message_name);
+                VERIFY_NOT_REACHED();
+            }
             VERIFY(!lexer.is_eof());
             consume_whitespace();
             parameter.name = lexer.consume_until([](char ch) { return isspace(ch) || ch == ',' || ch == ')'; });
@@ -177,10 +181,10 @@ Vector<Endpoint> parse(ByteBuffer const& file_contents)
         }
     };
 
-    auto parse_parameters = [&](Vector<Parameter>& storage) {
+    auto parse_parameters = [&](Vector<Parameter>& storage, StringView message_name) {
         for (;;) {
             consume_whitespace();
-            parse_parameter(storage);
+            parse_parameter(storage, message_name);
             consume_whitespace();
             if (lexer.consume_specific(','))
                 continue;
@@ -195,7 +199,7 @@ Vector<Endpoint> parse(ByteBuffer const& file_contents)
         message.name = lexer.consume_until([](char ch) { return isspace(ch) || ch == '('; });
         consume_whitespace();
         assert_specific('(');
-        parse_parameters(message.inputs);
+        parse_parameters(message.inputs, message.name);
         assert_specific(')');
         consume_whitespace();
         assert_specific('=');
@@ -212,7 +216,7 @@ Vector<Endpoint> parse(ByteBuffer const& file_contents)
 
         if (message.is_synchronous) {
             assert_specific('(');
-            parse_parameters(message.outputs);
+            parse_parameters(message.outputs, message.name);
             assert_specific(')');
         }
 

--- a/Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp
+++ b/Meta/Lagom/Tools/CodeGenerators/IPCCompiler/main.cpp
@@ -109,6 +109,37 @@ Vector<Endpoint> parse(ByteBuffer const& file_contents)
             lexer.ignore_until('\n');
     };
 
+    auto parse_parameter_type = [&]() {
+        ByteString parameter_type = lexer.consume_until([](char ch) { return ch == '<' || isspace(ch); });
+        if (lexer.peek() == '<') {
+            lexer.consume();
+
+            StringBuilder builder;
+            builder.append(parameter_type);
+            builder.append('<');
+            auto nesting_level = 1;
+            while (nesting_level > 0) {
+                auto inner_type = lexer.consume_until([](char ch) { return ch == '<' || ch == '>'; });
+                if (lexer.is_eof()) {
+                    warnln("Unexpected EOF when parsing parameter type");
+                    VERIFY_NOT_REACHED();
+                }
+                builder.append(inner_type);
+                if (lexer.peek() == '<') {
+                    nesting_level++;
+                } else if (lexer.peek() == '>') {
+                    nesting_level--;
+                }
+
+                builder.append(lexer.consume());
+            }
+
+            parameter_type = builder.to_byte_string();
+        }
+
+        return parameter_type;
+    };
+
     auto parse_parameter = [&](Vector<Parameter>& storage) {
         for (;;) {
             Parameter parameter;
@@ -133,14 +164,7 @@ Vector<Endpoint> parse(ByteBuffer const& file_contents)
                     consume_whitespace();
                 }
             }
-            // FIXME: This is not entirely correct. Types can have spaces, for example `HashMap<int, ByteString>`.
-            //        Maybe we should use LibCpp::Parser for parsing types.
-            parameter.type = lexer.consume_until([](char ch) { return isspace(ch); });
-            if (parameter.type.ends_with(',')) {
-                warnln("Parameter type '{}' looks invalid!", parameter.type);
-                warnln("Note that templates must not include spaces.");
-                VERIFY_NOT_REACHED();
-            }
+            parameter.type = parse_parameter_type();
             VERIFY(!lexer.is_eof());
             consume_whitespace();
             parameter.name = lexer.consume_until([](char ch) { return isspace(ch) || ch == ',' || ch == ')'; });

--- a/Userland/Services/Clipboard/ClipboardServer.ipc
+++ b/Userland/Services/Clipboard/ClipboardServer.ipc
@@ -2,6 +2,6 @@
 
 endpoint ClipboardServer
 {
-    get_clipboard_data() => (Core::AnonymousBuffer data, [UTF8] ByteString mime_type, HashMap<ByteString,ByteString> metadata)
-    set_clipboard_data(Core::AnonymousBuffer data, [UTF8] ByteString mime_type, HashMap<ByteString,ByteString> metadata) =|
+    get_clipboard_data() => (Core::AnonymousBuffer data, [UTF8] ByteString mime_type, HashMap<ByteString, ByteString> metadata)
+    set_clipboard_data(Core::AnonymousBuffer data, [UTF8] ByteString mime_type, HashMap<ByteString, ByteString> metadata) =|
 }

--- a/Userland/Services/RequestServer/RequestClient.ipc
+++ b/Userland/Services/RequestServer/RequestClient.ipc
@@ -5,7 +5,7 @@ endpoint RequestClient
     request_started(i32 request_id, IPC::File fd) =|
     request_progress(i32 request_id, Optional<u64> total_size, u64 downloaded_size) =|
     request_finished(i32 request_id, bool success, u64 total_size) =|
-    headers_became_available(i32 request_id, HashMap<ByteString,ByteString,CaseInsensitiveStringTraits> response_headers, Optional<u32> status_code) =|
+    headers_became_available(i32 request_id, HashMap<ByteString, ByteString, CaseInsensitiveStringTraits> response_headers, Optional<u32> status_code) =|
 
     // Websocket API
     // FIXME: See if this can be merged with the regular APIs

--- a/Userland/Services/RequestServer/RequestServer.ipc
+++ b/Userland/Services/RequestServer/RequestServer.ipc
@@ -6,14 +6,14 @@ endpoint RequestServer
     // Test if a specific protocol is supported, e.g "http"
     is_supported_protocol(ByteString protocol) => (bool supported)
 
-    start_request(i32 request_id, ByteString method, URL::URL url, HashMap<ByteString,ByteString> request_headers, ByteBuffer request_body, Core::ProxyData proxy_data) =|
+    start_request(i32 request_id, ByteString method, URL::URL url, HashMap<ByteString, ByteString> request_headers, ByteBuffer request_body, Core::ProxyData proxy_data) =|
     stop_request(i32 request_id) => (bool success)
     set_certificate(i32 request_id, ByteString certificate, ByteString key) => (bool success)
 
     ensure_connection(URL::URL url, ::RequestServer::CacheLevel cache_level) =|
 
     // Websocket Connection API
-    websocket_connect(URL::URL url, ByteString origin, Vector<ByteString> protocols, Vector<ByteString> extensions, HashMap<ByteString,ByteString> additional_request_headers) => (i32 connection_id)
+    websocket_connect(URL::URL url, ByteString origin, Vector<ByteString> protocols, Vector<ByteString> extensions, HashMap<ByteString, ByteString> additional_request_headers) => (i32 connection_id)
     websocket_ready_state(i32 connection_id) => (u32 ready_state)
     websocket_subprotocol_in_use(i32 connection_id) => (ByteString subprotocol_in_use)
     websocket_send(i32 connection_id, bool is_text, ByteBuffer data) =|

--- a/Userland/Services/WebContent/WebContentServer.ipc
+++ b/Userland/Services/WebContent/WebContentServer.ipc
@@ -70,7 +70,7 @@ endpoint WebContentServer
     set_content_filters(u64 page_id, Vector<String> filters) =|
     set_autoplay_allowed_on_all_websites(u64 page_id) =|
     set_autoplay_allowlist(u64 page_id, Vector<String> allowlist) =|
-    set_proxy_mappings(u64 page_id, Vector<ByteString> proxies, HashMap<ByteString,size_t> mappings) =|
+    set_proxy_mappings(u64 page_id, Vector<ByteString> proxies, HashMap<ByteString, size_t> mappings) =|
     set_preferred_color_scheme(u64 page_id, Web::CSS::PreferredColorScheme color_scheme) =|
     set_has_focus(u64 page_id, bool has_focus) =|
     set_is_scripting_enabled(u64 page_id, bool is_scripting_enabled) =|
@@ -79,8 +79,8 @@ endpoint WebContentServer
     set_window_position(u64 page_id, Web::DevicePixelPoint position) =|
     set_window_size(u64 page_id, Web::DevicePixelSize size) =|
 
-    get_local_storage_entries(u64 page_id) => (OrderedHashMap<String,String> entries)
-    get_session_storage_entries(u64 page_id) => (OrderedHashMap<String,String> entries)
+    get_local_storage_entries(u64 page_id) => (OrderedHashMap<String, String> entries)
+    get_session_storage_entries(u64 page_id) => (OrderedHashMap<String, String> entries)
 
     handle_file_return(u64 page_id, i32 error, Optional<IPC::File> file, i32 request_id) =|
 

--- a/Userland/Services/WindowServer/WindowClient.ipc
+++ b/Userland/Services/WindowServer/WindowClient.ipc
@@ -36,7 +36,7 @@ endpoint WindowClient
     drag_accepted() =|
     drag_cancelled() =|
 
-    drag_dropped(i32 window_id, Gfx::IntPoint mouse_position, [UTF8] ByteString text, HashMap<String,ByteBuffer> mime_data) =|
+    drag_dropped(i32 window_id, Gfx::IntPoint mouse_position, [UTF8] ByteString text, HashMap<String, ByteBuffer> mime_data) =|
 
     update_system_theme(Core::AnonymousBuffer theme_buffer) =|
     update_system_fonts(ByteString default_font_query, ByteString fixed_width_font_query, ByteString window_title_font_query) =|

--- a/Userland/Services/WindowServer/WindowServer.ipc
+++ b/Userland/Services/WindowServer/WindowServer.ipc
@@ -130,7 +130,7 @@ endpoint WindowServer
     set_window_cursor(i32 window_id, i32 cursor_type) =|
     set_window_custom_cursor(i32 window_id, Gfx::ShareableBitmap cursor) =|
 
-    start_drag([UTF8] ByteString text, HashMap<String,ByteBuffer> mime_data, Gfx::ShareableBitmap drag_bitmap) => (bool started)
+    start_drag([UTF8] ByteString text, HashMap<String, ByteBuffer> mime_data, Gfx::ShareableBitmap drag_bitmap) => (bool started)
     set_accepts_drag(bool accepts) =|
 
     set_system_theme(ByteString theme_path, [UTF8] ByteString theme_name, bool keep_desktop_background, Optional<ByteString> color_scheme_path) => (bool success)

